ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+				// use a cache of the most recent account state.
+				if offset == uint64(len(ao.deltas)) {
+					return macct.data.OnlineAccountData(rewardsProto, rewardsLevel), nil
+				}
+				// the account appears in the deltas, but we don't know if it appears in the
+				// delta range of [0..offset], so we'll need to check :
+				// Traverse the deltas backwards to ensure that later updates take
+				// priority if present.
+				for offset > 0 {
+					offset--
+					d, ok := ao.deltas[offset].GetData(addr)
+					if ok {
+						return d.OnlineAccountData(rewardsProto, rewardsLevel), nil
+					}
+				}
+			}
+		}
+
+		if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
+			return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+		}
+
+		ao.accountsMu.RUnlock()
+		needUnlock = false
+
+		// No updates of this account in the in-memory deltas; use on-disk DB.
+		// As an optimization, we avoid creating
+		// a separate transaction here, and directly use a prepared SQL query
+		// against the database.
+		persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+		if err != nil || persistedData.rowid == 0 {
+			// no such online account, return empty
+			return ledgercore.OnlineAccountData{}, err
+		}
+		// lookupOnlineHistory does not fetch the account db round because of the following observation:
+		// 1. ao.onlineAccountsCache update happens with ao.accountsMu taken below and in postCommit
+		// 2. If we started reading the history (lookupOnlineHistory)
+		//   1. before commitRound or while it is running => OK, read what is in DB and then add new entries in postCommit
+		//     * if commitRound deletes some history after, the cache has additional entries and updRound comparison gets a right value
+		//   2. after commitRound but before postCommit => OK, read full history, ignore the update from postCommit in writeFront's updRound comparison
+		//   3. after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here all the full history is loaded
+		persistedDataHistory, validThrough, err := ao.accountsq.lookupOnlineHistory(addr)
+		if err != nil || len(persistedDataHistory) == 0 {
+			return ledgercore.OnlineAccountData{}, err
+		}
+		// 3. If we fishied reading the history (lookupOnlineHistory)
+		//   1. db has not advanced (validThrough == currentDbRound) => OK
+		//   2. after commitRound but before postCommit (currentDeltaLen == len(ao.deltas)) => OK, the cache gets populated and postCommit updates the new entry
+		//   3. after commitRound and after postCommit => problem, postCommit does not add a new entry but the cache gets constructed by misses the latest entry, retry
+		// In order to resolve this lookupOnlineHistory returns dbRound value and determine what happened
+		// So handle 3.1 and 3.2 here and 3.3 below
+		ao.accountsMu.Lock()
+		if validThrough == currentDbRound || currentDbRound >= ao.cachedDBRoundOnline && currentDeltaLen == len(ao.deltas) {
```

Comment:
It may be worth putting this in a function so that you can close the lock with `defer` instead of closing it manually in 3 places.

Reply:
I respectfully disagree.

Reply:
I think I've even tried this before since I kind of fun of RAII and defers but at that iteration it just did the code even harder to understand.